### DISCUSSION TITLE
Keyboard compatibility questions

### DISCUSSION THREAD

**Discussion by wagnerck** [memtest86plus/memtest86plus#45] [General]
I'm an employee with [System76](https://system76.com/), and we make Linux systems (and a really nice keyboard). I can't _officially_ speak on behalf of the company, but a bunch of us in the tech support group are really looking forward to the new v6 memtest86+ with UEFI support etc, because it means we will no longer have to refer customers to the commercial memtest, which nags users to buy the Pro version and so on.

But for us to do that, we need to make sure that it will work on our hardware, and with the various keyboards that our customers will attach to our desktops. I've been doing some preliminary testing with keyboards and I'm finding mixed results with building an ISO from source, using it to make a memtest USB stick, and booting from it. Please don't take this as a complaint! I know that v6 isn't even in beta yet, and I want to do what I can to contribute here.

There's a few things i want to ask here before I continue with testing things. First, when v6 goes public beta later this month, will there be pre-built ISOs available for download, and if so, will there be different versions for the various `keyboard=` settings? And if not, what's the intended way of booting an ISO image on a USB stick with (for example) `keyboard=buggy-usb` set? It's pretty clear how to change the setting for an EFI file being loaded by an intermediate bootloader like `systemd-boot` (which works _great_, by the way), but it's not clear how to change that setting when booting from an ISO. Is this only available when building an ISO with GRUB?

Second, here at home I have a number of very different keyboards (old, new, mechanical, wireless, with integrated hubs, etc) and a number of different systems to connect them to, with radically different firmware types: BIOS, UEFI, Coreboot, and an Intel MacBook Air. Would doing testing on these for keyboard compatibility with the USB driver help the developers, or would it just be additional noise? My thought would be a kind of matrix showing how each keyboard works on each system, whether it needed `buggy-usb` or not, etc, but I don't want to go to all that work and have it not be useful, or do the testing using the pre-beta code when testing the public beta would be better.

I really appreciate the work you're all doing to give us the option of a fully open-source memtest again, and I hope that I can help in some way.

**Comment by martinwhitaker** [comment]
I don't know if @x86fr intends to make a pre-built ISO image available for beta testing, but if so, it would make sense to use GRUB to allow the various command line options to be exercised. The makefile already supports building a GRUB-based ISO.

Any testing of different keyboard and hardware combinations will be helpful. The USB drivers work with all the hardware I own, so any further development will be based on bug reports. Note there's a currently undocumented command line option, `noehci`, which provides a workaround if handover from the EHCI controller to its companion controllers fails (issue #6).

**Reply by x86fr** [reply]
> I don't know if @x86fr intends to make a pre-built ISO image available for beta testing, but if so, it would make sense to use GRUB to allow the various command line options to be exercised. The makefile already supports building a GRUB-based ISO.

Yes, prebuilt ISOs (generated with the standard Makefile) will be available with the first beta release.

**Comment by wagnerck** [comment]
I see that a `usbdebug` option was added today (https://github.com/memtest86plus/memtest86plus/commit/644a13c73038335162125a39ffe2882823bdc217), which seems like it'll be useful for those system+keyboard combinations that are misbehaving. I'll get started on the testing and share my results in an actual GitHub issue when I have something useful.

With some luck, I'll just have a list of devices that require `buggy-usb` and otherwise work fine. Crossing my fingers.

**Comment by x86fr** [comment]
Systems that requires `buggy-usb` flag are actually very uncommon. With the recent commits, the vast majority of USB keyboard should now work as expected without specific flags. If you find something wrong, please post in a new issue 1/ a screenshot of the blue detection screen (we've just added a 10 sec timer if no USB keyboard are detected), 2/ the motherboard model and 3/ keyboard used.

The first binary release (v6.00 beta 1) will be available within 2 weeks on memtest.org.

**Comment by wagnerck** [comment]
Issue created: https://github.com/memtest86plus/memtest86plus/issues/47

**Comment by wagnerck** [comment]
Copy-and-pasting from the issue since it's going to be closed:


This is the followup to the testing talked about this GitHub discussion (https://github.com/memtest86plus/memtest86plus/discussions/45). I built from source (at https://github.com/memtest86plus/memtest86plus/tree/66c6dfa3576f5720956848b7870fa04bf8da4c7d) with `make grub-iso`, created a USB stick, and ran a number of tests. This didn't take me as long as I expected as the problems were largely confined to one system (with multiple keyboards) and one keyboard (with multiple systems).

The five keyboards tested are:

* A Dell SK-8135 multimedia keuboard which has a USB1 hub built in (USB ID 413c:2010 https://web.archive.org/web/20111012163355/http://support.dell.com/support/edocs/acc/P76379/specs.htm)
* A System76 Launch mechanical keyboard with a USB3 hub built in, connected via a USB-C to USB-A cable (USB ID 3384:0001 https://tech-docs.system76.com/models/launch_1/README.html)
* A Logitech K400 wireless keyboard with touchpad, using a Logitech Unifying Receiver (USB ID 046d:c52b https://support.logi.com/hc/en-us/articles/360023306254-K400-Technical-Specifications)
* An HP KBAH21 keyboard, basically as generic an OEM keyboard as you can get (USB ID 0461:4e67 https://www.amazon.com/HP-801526-001-Keyboard-Galeras-wired/dp/B00XLVVEIM)
* A generic Chinese mechanical keyboard branded "Elecrat" (USB ID 258a:0190, original page has now disappeared from Amazon but it's basically this with slightly different branding, IDs as a "SINO WEALTH" device which is the microcontroller, I think: https://www.amazon.com/Water-Resistant-Backlit-Mechanical-Keyboard-Switches/dp/B01MRM4PB3)

The four systems used for testing are:

* A custom PC using an ASUS Z87-Expert motherboard, which has two different USB controllers built into it: the Intel XHCI controller from the Z87 chipset, and an additional ASMedia XHCI USB controller, but I've only listed it once because the results were the same regardless of which ports I used (UEFI, https://dlcdnets.asus.com/pub/ASUS/mb/LGA1150/Z87-EXPERT/E7833_Z87-EXPERT.pdf))
* A System76 Oryx Pro (model `orypx6`), using Coreboot firmware on a Intel HM470 chipset (UEFI, https://tech-docs.system76.com/models/oryp6/README.html)
* A 2013 MacBook Air (model `MacBookAir6,2`) with an Intel USB chipset I think? (Apple EFI, https://everymac.com/systems/apple/macbook-air/specs/macbook-air-core-i7-1.7-13-mid-2013-specs.html)
* An HP EliteBook 8460p with an Intel QM67 chipset (BIOS, http://h10032.www1.hp.com/ctg/Manual/c03253774.pdf)

The chart key for results:

* *Y* - works with the default boot mode
* *B* - works with the `keyboard=buggy-usb` option enabled
* *E* - works with the `noehci` option enabled
* *X* - does not work at all regardless of options

I am not doing testing with the `keyboard=legacy` mode, as this would often require end-users to change their UEFI configuration, but could do so later if needed.

|               | Dell    | Launch  | K400    | HP      | Elecrat |
|---------------|---------|---------|---------|---------|---------|
| Z87-Expert    |    E    |    E    |    E    |    Y    |    Y    |
| Oryx Pro      |    Y    |    Y    |    X    |    Y    |    Y*   |
| Macbook Air   |    Y    |    Y    |    X    |    Y    |    Y*   |
| HP 8460p      |    Y    |    Y    |    X    |    Y    |    Y    |

The Z87-Expert has a number of BIOS configuration options, including an "EHCI handoff" setting, which did not have any effect. The `noehci` option was required to get the K400, the Launch, and the Dell working with it, regardless of what BIOS settings were used. Enabling `buggy-usb` caused the Dell to fail to be detected at all; in all other cases the keyboards were detected (and show up on the blue screen as detected devices) but no keyboard input was accepted. I can provide screen photos but there didn't seem to be any useful information other than the keyboard being found.

The thing that all the keyboards that needed `noehci` on the Z87 have in common is that they're all running multiple devices through the one connection. The Logitech wireless receiver has a K400 keyboard and touchpad connecting to it, and the Dell and Launch both have actual USB hubs in them. The HP and the Elecrat are just straight keyboards with no extra bits other than muiltimedia controls. That seems like it might be important?

The K400 doesn't seem to work at all on most of these systems; the fact that it works on the Z87 with `noehci` but none of the other systems is weird, especially since the Z87 had problems with other keyboards as well. The K400 with the receiver always seemed to be recognized as a keyboard but the system would not accept any input from it.

The Elecrat would sometimes not work the first time, but work on a reboot; I would chalk that up to it being a no-name cheap keyboard that just misbehaves sometimes.

Your last comment in the discussion thread about `buggy-usb` being rarely required was correct. So the next question is, are the issues with the Z87 system and the Logitech receiver in general worth investigating further? The fact that the `noehci` option seems to fix the problems with the Z87 system suggests to me that it might be good to set that up as a menu option on the downloadable ISOs, assuming that they'll be using GRUB. That might be quicker than trying to debug the actual problem, which might be related to the Z87 firmware anyway. The K400 might just be a loss in terms of support, although I'd be interested to know if anyone else has better results with a different unifying receiver or keyboard.

---

Per @x86fr I'll be creating two new issues instead, one for the K400 and one for the Z87 issues.

**Comment by martinwhitaker** [comment]
I have a K260 keyboard with a unifying receiver ID 046d:c52e and a K270 keyboard with a unifying receiver ID 046d:c534, both of which work with memtest86+ on several hosts, variously using OHCI, EHCI, and XHCI controllers.

When you create the issue for the K400, could you show the output of `lsusb -v -d 046d:c52b`.

**Reply by wagnerck** [reply]
Providing the `lsusb` output was in fact the plan.

**Reply by wagnerck** [reply]
New issue is here: https://github.com/memtest86plus/memtest86plus/issues/48
